src/core/lib/gpr/alloc.cc

```diff
@@ -88,12 +133,58 @@ void* gpr_realloc(void* p, size_t size) {
 }
 
 void* gpr_malloc_aligned(size_t size, size_t alignment) {
-  GPR_ASSERT(((alignment - 1) & alignment) == 0);  // Must be power of 2.
-  size_t extra = alignment - 1 + sizeof(void*);
-  void* p = gpr_malloc(size + extra);
-  void** ret = (void**)(((uintptr_t)p + extra) & ~(alignment - 1));
-  ret[-1] = p;
-  return (void*)ret;
+#if defined(GPR_HAS_ALIGNED_ALLOC)
+  size = GPR_ROUND_UP_TO_SPECIFIED_SIZE(size, alignment);
+  void* ret = aligned_alloc(alignment, size);
+  GPR_ASSERT(ret != nullptr);
+  return ret;
+#elif defined(GPR_HAS_ALIGNED_MALLOC)
+  GPR_DEBUG_ASSERT(is_power_of_two(alignment));
```

Comment:
Why is this assertion being done in two of the three cases here?  Shouldn't it be done in all cases?